### PR TITLE
[WIP] TPU V1 Support Refactored

### PR DESCRIPTION
**Pull request by alexm-redhat**
This PR is a rebase and modification of @robertgshaw2-redhat original PR for TPU support in vLLM V1 from 2 months ago https://github.com/vllm-project/vllm/pull/10241

Currently, TPU attention kernel has no support for mixing prefills and decodes in the same scheduler iteration. As a result, this PR separates the requests to (1) prefills and (2) decodes, and executes each one of them separately. Google guys are working on a new TPU attention kernel that will allow mixing prefills and decodes, the moment it is ready, we will be able to remove the separation logic and unify the requests (which will also provide better performance).

Notes:

1. Verified correctness
2. Supports chunked prompt
3. Optimized python prompt/decode prepare loops
2. No TP > 1 support yet
4. Only greedy sampler for now
5. The TPU codepath in this PR is completely separated from the NVIDIA codepath. Specifically, the model_runner and worker are duplicated to avoid a dependency and complication with NVIDIA code. We decided to do it like this because we want to land the PR as soon as possible, and later work on the actual design of code sharing between different hardware backends.

Follow ups:
1. TP > 1 support
2. Prefix caching
3. Integration with Google's new rugged paged attention kernel
4. Optimize end-to-end, in particular, for large TP

### CODE REVIEW

**Comment by alexm-redhat** [comment]
@WoosukKwon @robertgshaw2-redhat @mgoin @vanbasten23 @liangfu @njhill the PR is refactored and ready for review.

**Comment by alexm-redhat** [comment]
Working on ensuring tests work and getting the compiler cache working

**Comment by bvrockwell** [comment]
thanks for the speedy turnaround @alexm-redhat ! @yaochengji if you could review please, that would be great!

**Comment by miladm** [comment]
cc @lsy323 @vanbasten23 @WoosukKwon to take a pass

File: vllm/platforms/cuda.py [hunk 1]
```diff
@@ -135,7 +135,7 @@ def check_and_update_config(cls, vllm_config: VllmConfig) -> None:
             else:
                 if envs.VLLM_USE_V1:
                     parallel_config.worker_cls = \
-                            "vllm.v1.worker.gpu_worker.Worker"
+                            "vllm.v1.worker.gpu_worker.GPUWorker"
```
Comment (WoosukKwon) [inline vllm/platforms/cuda.py:138]: I'm good with renaming this, but it seems not included in this PR?

**Review by WoosukKwon — COMMENTED** [review]
QQ: Can we remove V0 code (model runner, worker) for TPUs?

IIUC, this PR is a superset of V0, and we will only focus on the V1 branch.

File: vllm/v1/worker/tpu_worker.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+        self.model_runner.dummy_run(
+            runner_kv_caches,
+            num_tokens=1,
+            seq_len=self.scheduler_config.max_num_batched_tokens,
+            exec_mode=ExecutionMode.PREFILL,
+        )
+
+        # Synchronize before measuring the memory usage.
+        xm.wait_device_ops()
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_worker.py:153]: We should add `xm.mark_step()` before xm.wait_device_ops.

`xm.mark_step` is used to launch execution in device
`xm.wait_device_ops` is used to wait the launched execution in device to be finished.

File: tests/entrypoints/openai/test_accuracy.py [hunk 3]
```diff
@@ -67,14 +67,21 @@ def run_test(more_args):
                 ), f"Expected: {EXPECTED_VALUE} |  Measured: {measured_value}"
 
 
-@pytest.mark.skipif(not current_platform.is_cuda(),
-                    reason="V1 currently only supported on CUDA")
+@pytest.mark.skipif(not current_platform.is_cuda()
```
Comment (yaochengji) [inline tests/entrypoints/openai/test_accuracy.py:70]: Could we add all the important functionality test that vLLM TPU v1 can support?

File: vllm/v1/worker/tpu_worker.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+                self.cache_config.cache_dtype]
+
+        if self.model_config.trust_remote_code:
+            # note: lazy import to avoid importing torch before initializing
+            from vllm.utils import init_cached_hf_modules
+            init_cached_hf_modules()
+
+        # Torch profiler. Enabled and configured through env vars:
+        # VLLM_TORCH_PROFILER_DIR=/path/to/save/trace
+        if envs.VLLM_TORCH_PROFILER_DIR:
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_worker.py:69]: AFAIK, PyTorch/XLA cannot directly use the torch profiler

File: vllm/v1/worker/tpu_model_runner.py [hunk 5]
```diff
@@ -0,0 +1,1128 @@
...
+        # As a workaround, we use the xm's rank assignment only when loading
+        # the embedding weights.
+        xm_tp_rank = xr.global_ordinal()
+        with patch(
+                "vllm.model_executor.layers.vocab_parallel_embedding."
+                "get_tensor_model_parallel_rank",
+                return_value=xm_tp_rank):
+            model = get_model(vllm_config=self.vllm_config)
+        model = model.eval()
+        xm.wait_device_ops()
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:744]: We should add xm.mark_step() before xm.wait_device_ops.

xm.mark_step is used to launch execution in device
xm.wait_device_ops is used to wait the launched execution in device to be finished.

**Comment by bvrockwell** [comment]
> QQ: Can we remove V0 code (model runner, worker) for TPUs?
> 
> IIUC, this PR is a superset of V0, and we will only focus on the V1 branch.

there isn't feature parity between the two though right? Can we do this in the next PR when @vanbasten23 pipes in the new kernel and @alexm-redhat adds TP>1?

**Comment by alexm-redhat** [comment]
@WoosukKwon We will be able to remove V0 TPU in the next PR (after we get sampler and TP inside). But it is a good point.

File: examples/offline_inference/basic.py
Comment (WoosukKwon) [inline examples/offline_inference/basic.py:1]: Please keep this file unchanged!

File: tests/entrypoints/llm/test_accuracy.py [hunk 6]
```diff
@@ -21,10 +21,13 @@
 EXPECTED_VALUE = 0.58
 
 
-def run_test():
+def run_test(more_args=None):
     """Run the end to end accuracy test."""
 
-    model_args = f"pretrained={MODEL_NAME},max_model_len=2048"
+    model_args = f"pretrained={MODEL_NAME},max_model_len=4096"
```
Comment (WoosukKwon) [inline tests/entrypoints/llm/test_accuracy.py:27]: Just curious, why do we need to increase the model len?

File: vllm/v1/attention/backends/pallas.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+    def copy_blocks(
+        kv_caches: List[Tuple[torch.Tensor, torch.Tensor]],
+        src_to_dists: Tuple[torch.Tensor, torch.Tensor],
+    ) -> None:
+        src_indices, dst_indices = src_to_dists
+        for k_cache, v_cache in kv_caches:
+            torch.ops.xla.dynamo_set_buffer_donor_(k_cache, True)
+            k_cache[:, dst_indices] = k_cache[:, src_indices]
+            torch.ops.xla.dynamo_set_buffer_donor_(v_cache, True)
+            v_cache[:, dst_indices] = v_cache[:, src_indices]
```
Comment (WoosukKwon) [inline vllm/v1/attention/backends/pallas.py:61]: btw, V1 doesn't use this method.

**Review by WoosukKwon — APPROVED** [review]
As we discussed offline, I'm good with merging the PR to unblock other efforts, despite a lot of code duplication (that I asked) here. We will come back and refactor this out. Specifically,

1. We should drop the V0 code once the V1 code path replaces the functionality.
2. We should refactor out the TPU worker with the recently introduced `WorkerBase`.
3. We should refactor out the common logic between V1 model runners. (I will explore this.)

Thanks for the work! Please make sure the `basic.py` file is unchanged.

File: tests/entrypoints/llm/test_accuracy.py (hunk 6 above)
Comment (alexm-redhat) [reply tests/entrypoints/llm/test_accuracy.py:27]: There is some compilation related issue with 2048 that I was able to fully resolve. I will dig more into this in the next PR.

File: vllm/v1/attention/backends/pallas.py (hunk 7 above)
Comment (alexm-redhat) [reply vllm/v1/attention/backends/pallas.py:61]: This file is going to change, @vanbasten23 is working on the integration with the new kernel

**Comment by alexm-redhat** [comment]
@WoosukKwon thanks for the detailed review. I have reverted basic.py, hope tests will be green now.
